python/cuml/experimental/cluster/hdbscan.pyx

```diff
@@ -0,0 +1,550 @@
+        cdef uintptr_t sizes_ptr = self.sizes_.ptr
+        cdef uintptr_t lambdas_ptr = self.lambdas_.ptr
+        cdef uintptr_t probabilities_ptr = self.probabilities_.ptr
+        cdef uintptr_t mst_src_ptr = self.mst_src_.ptr
+        cdef uintptr_t mst_dst_ptr = self.mst_dst_.ptr
+        cdef uintptr_t mst_weights_ptr = self.mst_weights_.ptr
+
+        # If calling fit a second time, release
+        # any memory owned from previous trainings
+        delete_hdbscan_output(self)
+
+        cdef hdbscan_output *linkage_output = new hdbscan_output(
+            handle_[0], n_rows,
+            <int*>labels_ptr,
+            <float*>probabilities_ptr,
+            <int*>children_ptr,
+            <int*>sizes_ptr,
+            <float*>lambdas_ptr,
+            <int*>mst_src_ptr,
+            <int*>mst_dst_ptr,
+            <float*>mst_weights_ptr)
+
+        self.hdbscan_output_ = <size_t>linkage_output
+
+        cdef HDBSCANParams params
+        params.k = self.n_neighbors
+        params.min_samples = self.min_samples
+        # params.alpha = self.alpha
+        params.min_cluster_size = self.min_cluster_size
+        params.max_cluster_size = self.max_cluster_size
+        params.cluster_selection_epsilon = self.cluster_selection_epsilon
+        params.allow_single_cluster = self.allow_single_cluster
+
+        if self.cluster_selection_method == 'eom':
+            params.cluster_selection_method = CLUSTER_SELECTION_METHOD.EOM
+        elif self.cluster_selection_method == 'leaf':
+            params.cluster_selection_method = CLUSTER_SELECTION_METHOD.LEAF
+        else:
+            raise ValueError("Cluster selection method not supported. "
+                             "Must one of {'eom', 'leaf'}")
+
+        cdef DistanceType metric
+        if self.metric in _metrics_mapping:
+            metric = _metrics_mapping[self.metric]
+        else:
+            raise ValueError("'affinity' %s not supported." % self.affinity)
+
+        if self.connectivity == 'knn':
+            hdbscan(handle_[0],
+                    <float*>input_ptr,
+                    <int> n_rows,
+                    <int> n_cols,
+                    <DistanceType> metric,
+                    params,
+                    deref(linkage_output))
+        else:
+            raise ValueError("'connectivity' can only be one of "
+                             "{'knn', 'pairwise'}")
+
+        self.handle.sync()
+
+        self.fit_called_ = True
+
+        self._construct_output_attributes()
+
+        try:
+
+            from hdbscan.plots import SingleLinkageTree
+
+            if self.gen_min_span_tree_:
+                self._minimum_spanning_tree = \
+                    self._build_minimum_spanning_tree(X_m)
+
+            self._condensed_tree = self._build_condensed_tree()
+            self._single_linkage_tree = self._build_single_linkage_tree()
+
+        except Exception as e:
+            logger.warn("hdbscan must be installed to use plots")
```

Comment:
I think it might be a good idea to have a parameter to not generate these if the user doesn't want (even if they have `hdbscan` installed) to avoid memory transfers and whatnot, what do you think? They could also be part of another method that the user can call to build explicitly.

Reply:
I totally agree. I moved these in here last minute because I couldn't get the `@property` to work. I'd personally prefer to be able to lazy-load these values if we could get that to work.